lib/spack/spack/test/detection.py

```diff
@@ -0,0 +1,208 @@
+    assert path_to_exe[cmake_path1] == 'cmake'
+
+
+def test_find_external_merge(mutable_config, mutable_mock_repo):
+    """Check that 'spack find external' doesn't overwrite an existing spec
+    entry in packages.yaml.
+    """
+    pkgs_cfg_init = {
+        'find-externals1': {
+            'externals': [{
+                'spec': 'find-externals1@1.1',
+                'prefix': '/preexisting-prefix/'
+            }],
+            'buildable': False
+        }
+    }
+
+    mutable_config.update_config('packages', pkgs_cfg_init)
+    entries = [
+        spack.detection.DetectedPackage(
+            spack.spec.Spec.from_detection('find-externals1@1.1'), '/x/y1/'
+        ),
+        spack.detection.DetectedPackage(
+            spack.spec.Spec.from_detection('find-externals1@1.2'), '/x/y2/'
+        )
+    ]
+    pkg_to_entries = {'find-externals1': entries}
+    scope = spack.config.default_modify_scope('packages')
+    spack.detection.update_configuration(pkg_to_entries, scope=scope, buildable=True)
+
+    pkgs_cfg = spack.config.get('packages')
+    pkg_cfg = pkgs_cfg['find-externals1']
+    pkg_externals = pkg_cfg['externals']
+
+    assert {'spec': 'find-externals1@1.1',
+            'prefix': '/preexisting-prefix/'} in pkg_externals
+    assert {'spec': 'find-externals1@1.2',
+            'prefix': '/x/y2/'} in pkg_externals
+
+
+def candidate_packages():
+    """Return the list of packages with a corresponding
+    detection_test.yaml file.
+    """
+    # Directories where we have repositories
+    repo_dirs = [spack.util.path.canonicalize_path(x)
+                 for x in spack.config.get('repos')]
+
+    # Compute which files need to be tested
+    to_be_tested = []
+    for repo_dir in repo_dirs:
+        pattern = os.path.join(
+            repo_dir, 'packages', '**', 'detection_test.yaml'
+        )
+        pkgs_with_tests = [os.path.basename(os.path.dirname(x))
+                           for x in glob.glob(pattern)]
+        to_be_tested.extend(pkgs_with_tests)
+
+    return to_be_tested
+
+
+@pytest.mark.detection
+@pytest.mark.parametrize('package_name', candidate_packages())
+def test_package_detection(mock_executable, package_name):
+    def detection_tests_for(pkg):
+        pkg_dir = os.path.dirname(
+            spack.repo.path.filename_for_package_name(pkg)
+        )
+        detection_data = os.path.join(pkg_dir, 'detection_test.yaml')
+        with open(detection_data) as f:
+            return spack.util.spack_yaml.load(f)
+
+    @contextlib.contextmanager
+    def setup_test_layout(layout):
+        hints, to_be_removed = set(), []
+        for binary in layout:
+            exe = mock_executable(
+                binary['name'], binary['output'], subdir=binary['subdir']
+            )
+            to_be_removed.append(exe)
+            hints.add(os.path.dirname(str(exe)))
+
+        yield list(hints)
+
+        for exe in to_be_removed:
+            os.unlink(exe)
```

Comment:
I think this can move out from being nested as well, but it will require some changes to how it calls mock_executable, so that might not be worth it.

Reply:
Outdated